### PR TITLE
Add prefix sum function to C4 library

### PR DESCRIPTION
**Pull request by alexrlongne**
+ The MPI scan function calculates a prefix sum (often used in determining offsets in parallel problems) without forming a n_proc length array on every node.
+ Templated prefix_sum function was added to the "blocking" section of the C4 library
+ A section for prefix_sum was added to the "tstReduction" test
* Purpose of Pull Request
  * Address Jayenne #892 (https://rtt.lanl.gov/redmine/issues/892)
* [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [ ] Travis CI checks pass
  * [x] Code coverage does not decrease
  * [x] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Toss2 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco) (regression broken after slurm install).
  * [x] Code reviewed/approved, sufficient DbC checks, testing, documentation

### CODE REVIEW

File: src/c4/test/tstReduction.cc [hunk 1]
```diff
@@ -190,11 +191,72 @@ void array_reduction(rtt_dsxx::UnitTest &ut) {
...
+
+  if (!soft_equiv(xdbl_prefix_sum, dbl_answer))
+    ITFAILS;
+
+  if (ut.numFails == 0)
+    PASSMSG("Prefix sum ok.");
+  return;
+}
+
+//---------------------------------------------------------------------------//
```
Comment (KineticTheory) [inline src/c4/test/tstReduction.cc]: Please add tests for uint32_t and float.  These should be trivial to add.

File: src/c4/C4_MPI.t.hh [hunk 2]
```diff
@@ -353,6 +353,15 @@ template <typename T> DLL_PUBLIC_c4 void global_max(T *x, int n) {
...
 
+//---------------------------------------------------------------------------//
+
+template <typename T> DLL_PUBLIC_c4 T prefix_sum(T &node_value) {
+  T prefix_sum = 0;
+  MPI_Scan(&node_value, &prefix_sum, 1, MPI_Traits<T>::element_type(), MPI_SUM,
+           communicator);
+  return prefix_sum;
+}
+
```
Comment (KineticTheory) [inline src/c4/C4_MPI.t.hh]: You need to add an equivalent function to C4_Serial.hh.

File: src/c4/C4_Functions.hh [hunk 3]
```diff
@@ -466,6 +466,18 @@ DLL_PUBLIC_c4 bool isScalar();
...
 DLL_PUBLIC_c4 std::string get_processor_name();
 
+//---------------------------------------------------------------------------//
+// prefix_sum
+//---------------------------------------------------------------------------//
+/*!
+ * \brief Return the value of the prefix sum at this processor.
+ *
+ * \param node_value
+ * Current node's value of variable to be prefix summed
```
Comment (KineticTheory) [inline src/c4/C4_Functions.hh]: Put the description on the same line as the `\param` or check that this produces the correct output (make autodoc).  See examples at https://www.stack.nl/~dimitri/doxygen/manual/docblocks.html.

File: src/c4/C4_MPI.t.hh (hunk 2 above)
Comment (alexrlongne) [inline src/c4/C4_MPI.t.hh]: prefix_sum now has a serial implementation and I tested and updated the C4_SERIAL build (with your help)

File: src/c4/C4_Functions.hh (hunk 3 above)
Comment (alexrlongne) [inline src/c4/C4_Functions.hh]: I've moved the description to the same line, as is done in the examples.

File: src/c4/C4_MPI.t.hh (hunk 2 above)
Comment (KineticTheory) [inline src/c4/C4_MPI.t.hh]: Are there any DbC statements that can be inserted into this function?  Is there a valid range for node_value?

**Pull request merged**